_posts/2016-04-26-distributed-game-of-life-in-elixir.md

```diff
@@ -0,0 +1,1065 @@
+{% endhighlight %}
+
+You can find [more patterns in modules here](https://github.com/BeyondScheme/elixir-game_of_life/tree/master/lib/game_of_life/patterns).
+
+# Run game across multiple nodes
+
+Now it's time to run our game. The full [source code can be found here](https://github.com/BeyondScheme/elixir-game_of_life).
+
+Let's run first node where the `GameOfLife.BoardServer` will be running.
+
+{% highlight plain %}
+$ iex --sname node1 -S mix
+Erlang/OTP 18 [erts-7.3] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]
+Interactive Elixir (1.2.4) - press Ctrl+C to exit (type h() ENTER for help)
+
+16:54:08.554 [info]  Started Elixir.GameOfLife.BoardServer master
+
+iex(node1@Artur)1> GameOfLife.BoardServer.start_game
+:game_started
+
+iex(node1@Artur)2> GameOfLife.GamePrinter.start_printing_board
+:printing_started
+{% endhighlight %}
+
+In another terminal window you can start second node. We will connect it with the first node.
+
+{% highlight plain %}
+$ iex --sname node2 -S mix
+Erlang/OTP 18 [erts-7.3] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]
+Interactive Elixir (1.2.4) - press Ctrl+C to exit (type h() ENTER for help)
+
+16:55:17.395 [info]  Started Elixir.GameOfLife.BoardServer master
+
+iex(node2@Artur)1> Node.connect :node1@Artur
+true
+16:55:17.691 [info]  Started Elixir.GameOfLife.BoardServer slave
+
+iex(node2@Artur)2> Node.list
+[:node1@Artur]
+
+iex(node2@Artur)3> Node.self
+:node2@Artur
+
+iex(node2@Artur)4> GameOfLife.Patterns.Guns.gosper_glider |> GameOfLife.BoardServer.add_cells
+[{24, 8}, {22, 7}, {24, 7}, {12, 6}, {13, 6}, {20, 6}, {21, 6}, {34, 6},
+ {35, 6}, {11, 5}, {15, 5}, {20, 5}, {21, 5}, {34, 5}, {35, 5}, {0, 4}, {1, 4},
+ {10, 4}, {16, 4}, {20, 4}, {21, 4}, {0, 3}, {1, 3}, {10, 3}, {14, 3}, {16, 3},
+ {17, 3}, {22, 3}, {24, 3}, {10, 2}, {16, 2}, {24, 2}, {11, 1}, {15, 1},
+ {12, 0}, {13, 0}]
+{% endhighlight %}
+
+Both nodes are executing calculation to determine a new state for living cells. You can run game also across diferrent servers in the network. You can do that like that:
+
+{% highlight plain %}
+# start node1
+$ iex --name node1@192.168.0.101 --cookie "token_for_cluster" -S mix
+
+# start node2 on another server
+$ iex --name node2@192.168.0.102 --cookie "token_for_cluster" -S mix
+iex> Node.connect :"node1@192.168.0.101"
+true
+{% endhighlight %}
+
+You could already see how the game works in the demo at the beginning of the article. You can try it on your own machine, just clone the [repository](https://github.com/BeyondScheme/elixir-game_of_life).
+
+# Summary
+
+Finnally we managed to get to the end. It was a pretty long road but we have a working game, distributed across nodes. We learned how to write GenServer, how to use Agents and how we can split processes across nodes with TaskSupervisor. You also saw examples of tests in Elixir and how to use DocTest. Now you know how to connect nodes into the cluster.
```

Comment:
We learned how to write GenServer,  use Agents. split processes across nodes with TaskSupervisor and connect nodes into the cluster. You also saw examples of tests in Elixir and how to use DocTest.